Homework #6

BolzmanBrain opened the pull request:
- now using Hibernate to access DB
- added BookComment

BolzmanBrain committed:
Homework #6

- now using Hibernate to access DB
- added BookComment

BolzmanBrain committed:
Homework #6

- added read only for @Transaction of methods which only select data

kataus commented on homework6/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,70 @@
+    @Column(name = "name", nullable = false)
+    private final String name;
+
+    @OneToOne(fetch = FetchType.LAZY)
```
Для каждой книги нужно будет создавать своего уникального автора и свой уникальный жанр?

BolzmanBrain replied:
Поменял вид связи на @ManyToOne.

kataus commented on homework6/src/main/java/ru/otus/spring/domain/BookComment.java:
```diff
@@ -0,0 +1,40 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private final Long id;
+
+    @Column(name = "text", nullable = false, length = 8000)
```
nullable, length - это инструкции для создания сприптов для базы руками JPA. Больше они никак не используются

BolzmanBrain replied:
Убрал.

kataus commented on homework6/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,70 @@
+@Entity
+@Table(name = "books_tbl")
+@Getter
+@Setter
```
А почему не Data? Где-то так, где-то так. Смущает такой разнобой

BolzmanBrain replied:
Нам Александр на лекции говорил, что из-за Data могут быть проблемы при работе Hibernate с классами, у которых есть LAZY-поля, потому что Data перегружает методы Object.
Но поменял, проблем не возникло.

kataus commented on homework6/src/main/java/ru/otus/spring/domain/BookComment.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public String convertToString() {
+        return String.format("BookComment(id = %d, text = %s, idBook = %d)",
```
А почему бы просто toString не переопределить и не мучаться с отдельным интерфейсом?

BolzmanBrain replied:
Думал, что могут возникнуть при работе с сущностями, у которых есть LAZY-поля. Убрал.

kataus commented on homework6/src/main/java/ru/otus/spring/exceptions/UniqueKeyViolatedException.java:
```diff
@@ -0,0 +1,10 @@
+package ru.otus.spring.exceptions;
+
+public class UniqueKeyViolatedException extends RuntimeException {
```
Удобно использовать своё единое корневое исключение. Тогда при обработке можно будет легко отделить наши и не наши исключения

BolzmanBrain replied:
Сделал одно исключение.

kataus commented on homework6/src/main/java/ru/otus/spring/repository/AuthorRepositoryJpa.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Objects;
+import java.util.Optional;
+
+@Repository
```
Что тут используется от аннотации @repository? Она несколько больше чем @Component или @Service. Если она появляется -то мы ожидаем использование её особенностей

BolzmanBrain replied:
Пока не ознакомился с дополнительным функционалом, который она предоставляет.
Но вообще, в теории, если бы у меня использовался AOP (знаю, что такая необходимость возникает редко), и нужно было бы повесить Advice на методы репозитория, то это было бы полезно.

Поменял на Service.

kataus commented on homework6/src/main/java/ru/otus/spring/repository/BookCommentRepositoryJpa.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public List<BookComment> getByBookId(long idBook) {
```
Давайте удалим из всех dao ручное получение комментриев по книге. Решим задачу методами ORM с помощью связей в сущностях. Решение получится на слое бизнес-логики

BolzmanBrain replied:
Сделал. Действительно, не подумал, что разумнее просто выгрузить книгу и воспользоваться магией ORM :)

kataus commented on homework6/src/main/java/ru/otus/spring/shell/AuthorShell.java:
```diff
@@ -0,0 +1,98 @@
+public class AuthorShell {
+    private final AuthorRepository authorRepository;
+
+    @Transactional(readOnly = true)
```
Транзакция - не задача уровня представления

BolzmanBrain replied:
Перенёс на уровень сервисов.

kataus commented on homework6/src/main/java/ru/otus/spring/shell/AuthorShell.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private String outputAllAuthors() {
```
Принят следующий порядок методов в классе: конструкторы, public, protected, private. Мотивация - общение с внешним миром видим сразу, а за внутренними деталями можно и проскролить класс. При такой привычке просто не ожидаешь публичные методы после приватных

BolzmanBrain replied:
Убрал private-методы в конец классов.

kataus commented on homework6/src/main/java/ru/otus/spring/shell/BookCommentShell.java:
```diff
@@ -0,0 +1,104 @@
+    public String insert(@ShellOption(value = {"--book-id","-b"}) Long idBook,
+                         @ShellOption(value = {"--text","-t"}) String text) {
+        try {
+            Book book = bookRepository.getById(idBook).orElseThrow();
```
Как много логики, а казалось что класс отвечает за взаимодействие с пользователем

BolzmanBrain replied:
Добавил уровень сервисов.

kataus commented on homework6/src/main/java/ru/otus/spring/shell/BookCommentShell.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    @Transactional(readOnly = true)
```
Какая тут польза от транзакции ?

BolzmanBrain replied:
Убрал транзакции там где можно обойтись без них.

Теперь они остались только там где происходит изменение данных и инициализация LAZY-полей.

BolzmanBrain committed:
Homework #6:

- fixes after receiving feedback
- created service layer
- now repository.delete takes object parameter
- removed @Transactional from shell to service layer
- removed @Transactional when it isn't nescessary (selects without loading lazy fields)
- removed RepresentableAsString interface
- removed 'nullable' and 'length' params in columns
- fixed order of methods in class (public, protected, private)

BolzmanBrain committed:
Homework 6:

- removed debug sout
- fixed updating books (it was deleting comments due to Casscade.ALL)

kataus commented on homework6/src/main/java/ru/otus/spring/domain/Author.java:
```diff
@@ -0,0 +1,34 @@
+    @Column(name = "name", unique = true)
+    private final String name;
+
+    public Author() {
```
Можно добавить @NoArgsContructor и не делать вручную

BolzmanBrain replied:
Хотел именно так и сделать, но почему-то в таком случае появляется ошибка: "java: variable id might not have been initialized"

kataus commented on homework6/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,72 @@
+    private final String name;
+
+    @Fetch(FetchMode.JOIN)
+    @ManyToOne(fetch = FetchType.LAZY)
```
а есть ли смысл в книгах без автора и жанра? тут не просто так EAGER по умолчанию

BolzmanBrain replied:
Исправил

kataus commented on homework6/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,72 @@
+        this.genre = new Genre();
+    }
+
+    public Book(Long id, String name, Author author, Genre genre) {
```
стоит единый стиль держать, или ручные конструкторы или ломбок

BolzmanBrain replied:
Не знал, что @RequiredArgsConstructor учитывает, что не нужно создавать конструктор с параметром, который инициализируется в теле класса. Исправил.

kataus commented on homework6/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,72 @@
+    @OneToMany(cascade = {CascadeType.PERSIST, CascadeType.REMOVE},
+            mappedBy = "book",
+            orphanRemoval = true,
+            fetch = FetchType.LAZY)
```
дефолтные значения не стоит повторять. а то смущает, приходится искать причину зачем это сделано

BolzmanBrain replied:
Согласимся. Convention over configuration :)

kataus commented on homework6/src/main/java/ru/otus/spring/repository/BookRepositoryJpa.java:
```diff
@@ -0,0 +1,58 @@
+    private final EntityManager em;
+
+    // transactional is needed for correct retrieval of LAZY-field "comments"
+    @Transactional(readOnly = true)
```
почему как раз EAGER. При каждом запросе всё равно приходится мучаться чтобы корректно авторы и жанры появились

BolzmanBrain replied:
Кажется, я [догадываюсь](https://www.youtube.com/watch?v=SR4aWoH2DN4&ab_channel=MrSkaso4nik) в чём дело)

kataus commented on homework6/src/main/java/ru/otus/spring/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    private final GenreRepository genreRepository;
+    private final BookRepository bookRepository;
+
+    // @Transactional is needed for initialization of lazy fields
```
и тут и в dao... Сложно как-то вышло, много лишнего
